### ISSUE TITLE
implement dbAppendTable

### ISSUE THREAD

**Issue by r2evans** [rstudio/pool#85]
Perhaps related to #32, number 3, inheriting from `DBI*`. There are several functions working (`DBI::dbGetQuery`, et al), but `DBI::dbAppendTable` does not work for signature "Pool".

**Comment by JasperSch** [comment]
I assume some of the functions (dbSendQuery, dbAppendTable,...) have not been implemented because of this reason in the documentation:

_You cannot perform SQL transaction using a Pool object directly (because that would imply keeping a connection open and not knowing when to return it back to the pool.)_

Right now I include functions of the following kind within my packages:

```
#' dbAppendTable for Pool objects
#' 
#' @param conn Pool object as returned by  \link[pool]{dbPool}
#' @param ... arguments passed to \link[DBI]{dbAppendTable}
#' @importFrom DBI dbAppendTable
#' @importFrom pool poolCheckout poolReturn
#' @export
dbAppendTable.Pool <- function(conn, ...){  
  myConnection <- poolCheckout(conn)
  on.exit(poolReturn(myConnection))  
  DBI::dbAppendTable(conn = myConnection, ...)
}

```

I'm however curious to know if there are any pitfalls to this approach and possible reasons for not including these functions within in the pool package.

**Comment by JasperSch** [comment]
Related to #88

**Comment by r2evans** [comment]
I don't disagree that `Pool` itself has no concept of actual SQL dialects, but ... it's often recommended to load `DBI` but not load the driver package itself (e.g., `library(odbc)`), so the methods are not necessarily registered. I don't know if there's a better way to handle it other than loading the driver package as well. I think #88's recommendations 1 and 2 are *wrong*: the intent of a parent `DBI` package is so that code *just works* with various drivers, and `pool` is an enabler. Having a requirement to change the functions used based on the driver is going down the wrong road. The third recommendation (transactions) makes sense in many ways anyway, but still ... I think we need to find a way to get `dbAppendTable` to work when operating on a `pool` connection, otherwise we're fracturing the code a little.

**Comment by jcheng5** [comment]
`dbSendQuery` is implemented. I think `dbAppendTable` is only not implemented because it didn't exist at the time of pool's release. I don't know about `dbSendStatement`, maybe just an oversight. I'm planning to submit pool to CRAN next week, I can throw in the missing methods.

**Comment by JasperSch** [comment]
@jcheng5  Thank you very much!

**Comment by jcheng5** [comment]
Sorry, looking at the code now, `dbSendQuery` is implemented but only to throw an error. It and `dbSendStatement` both return live ResultSet objects, which is fundamentally incompatible with pool when you use it in the mode of "pretend this pool is a DBI connection"; by the time the caller gets around to using the ResultSet object, the underlying connection may have been used for something else or possibly even closed.

For what it's worth, my initial vision of pool _required_ you to checkout/return connections from the pool explicitly every time. This is in keeping with how every other DB connection pooling library works in every other language, that I have personally encountered. However, this proved to be excessive boilerplate for the 99% of uses by R users where the operation they're performing is compatible with "poolCheckout, do the thing, poolReturn, return result". So where possible, we implement the DBI methods as a convenience.

Looking at the README, I see that we say "Essentially, you shouldn't feel the difference, with the exception of creating and closing a Pool object (as opposed to connecting and disconnecting a DBIConnection object)" which is probably an overreach. It is actually true for the vast majority of Shiny apps and Plumber APIs (where we envisioned pool being the most useful) but is not the way I personally think of this package.

**Comment by r2evans** [comment]
I agree that the behavior of pool should be consistent with other languages, and the way you explain `dbSendStatement`'s behavior makes sense to me. In essence, anything that allows (requires) the user to do something with *this resultset* can easily break, since there's no guarantee that it's the same connection (and therefore the old resultset will have been discarded/garbage-collected).

So this behavior makes sense for two-step functions, i.e., Send then Fetch (or similar). What do you think about `dbAppendTable`? From the user-perspective, it's a one-step function, which means there is reasonable assurance that the connection is unchanged. Since `DBI::dbAppendTable` merely calls `dbExecute` (https://github.com/r-dbi/DBI/blob/306e3f1c973b6208cee3b9996fcf3187c5d6c5c4/R/table-insert.R#L150), I suspect it could be a straight-forward method. Am I missing steps that would complicate this?

**Comment by jcheng5** [comment]
Yup, `dbAppendTable` is fine, I'm about to push a branch with it.

**Comment by jcheng5** [comment]
Fixed in #109. Thanks!
